### ISSUE TITLE
Refactoring torchtext.datasets (and _RawIterableDataset) to work with torch datapipes (and other things) by changing _RawIterableDataset's __iter__ .

### ISSUE THREAD

**Issue by bentrevett** [pytorch/text#1243]
## Problem

PyTorch `IterableDataset`s should implement an `__iter__` method which should call a function that creates a fresh iterator for the dataset. The new torchtext datasets use a `_RawIterableDataset`, which inherits from PyTorch's `IterableDataset`, but its `__iter__` method is only implemented as `return self`. This means if you can only iterate over a `_RawIterableDataset` dataset once before it's empty and can't be iterated over again. I am not sure if is intended or not -- or if I am not understanding the abstractions correctly.

## Datapipes

On another note, PyTorch is adding a `DataPipe` class which reduces the responsbilities of the `DataLoader` abstraction. This is currently only available in the nightly version of PyTorch. See the RFC [here](https://github.com/pytorch/pytorch/issues/49440), and the current datapipe code [here](https://github.com/pytorch/pytorch/tree/master/torch/utils/data/datapipes).

Here's an example of using datapipes with the latest version of torchtext (not included is using a Vocab to numericalize the examples):

```python
from torchtext.datasets import IMDB
from torchtext.data.utils import get_tokenizer
import torch.utils.data.datapipes as datapipes

train_data = IMDB(root='data', split='train')

tokenizer = get_tokenizer('basic_english')

def fn(example):
    label, text = example
    tokens = tokenizer(text)
    return label, tokens

def collate_fn(batch):
    label, text = zip(*batch)
    return label, text

# applies tokenization
pipe = datapipes.iter.Map(train_data, fn=fn)
# shuffles dataset
pipe = datapipes.iter.Shuffle(pipe, buffer_size=100)
# buckets batches based on length
pipe = datapipes.iter.BucketBatch(pipe, batch_size=5, bucket_size_mul=100, sort_key=lambda x: len(x[-1]))
# collates batches together
pipe = datapipes.iter.Collate(pipe, collate_fn=collate_fn)
```

You can iterate over the `pipe` to get the `(label, tokens)`, but again because the `_RawIterableDataset`'s `__iter__` method only returns `self`, you can only iterate over it once.

## Potential Solution

One idea is to have each dataset implement their own custom `__iter__` function. One solution with the IMDB dataset: 

```python
@_add_docstring_header(num_lines=NUM_LINES, num_classes=2)
@_wrap_split_argument(('train', 'test'))
class IMDB(torch.utils.data.IterableDataset):

    def __init__(self, root, split):
        super(IMDB, self).__init__()
        self.root = root
        self.split = split
        dataset_tar = download_from_url(URL, root=self.root,
                                    hash_value=MD5, hash_type='md5')
        self.dataset_files = extract_archive(dataset_tar)

    def __iter__(self):
        for fname in self.dataset_files:
            if 'urls' in fname:
                continue
            elif self.split in fname and ('pos' in fname or 'neg' in fname):
                with io.open(fname, encoding="utf8") as f:
                    label = 'pos' if 'pos' in fname else 'neg'
                    yield label, f.read()
```

The problem is that each dataset needs their own unique `__iter__` function as all datasets are completely different and that this gets rid of the `_RawIterableDataset`, which seems like quite a useful abstraction.

Another option is to not pass the iterator directly to `_RawIterableDataset` but instead pass a partial function which can be called to get the iterator:

```python
@_add_docstring_header(num_lines=NUM_LINES, num_classes=2)
@_wrap_split_argument(('train', 'test'))
def IMDB(root, split):
    def generate_imdb_data(key, extracted_files):
        for fname in extracted_files:
            if 'urls' in fname:
                continue
            elif key in fname and ('pos' in fname or 'neg' in fname):
                with io.open(fname, encoding="utf8") as f:
                    label = 'pos' if 'pos' in fname else 'neg'
                    yield label, f.read()
    dataset_tar = download_from_url(URL, root=root,
                                    hash_value=MD5, hash_type='md5')
    extracted_files = extract_archive(dataset_tar)
    # iterator is now a partial function called iterator_fn
    iterator_fn = functools.partial(generate_imdb_data, split, extracted_files)
    return _RawTextIterableDataset("IMDB", NUM_LINES[split], iterator_fn)

class _RawTextIterableDataset(torch.utils.data.IterableDataset):
    """Defines an abstraction for raw text iterable datasets.
    """

    def __init__(self, description, num_lines, iterator_fn):
        """Initiate the dataset abstraction.
        """
        super(_RawTextIterableDataset, self).__init__()
        self.description = description
        self._iterator_fn = iterator_fn
        self.num_lines = full_num_lines

    def __iter__(self):
        # calls the partial function on iteration
        return self._iterator_fn() 

    def __len__(self):
        return self.num_lines

    def __str__(self):
        return self.description
```

There are probably some downsides to this approach, but none are coming to mind at the moment so would be interested to know what they are, or if there is a way to "reset" the `IterableDataset` which I'm unaware of.

**Comment by zhangguanheng66** [comment]
Thanks @bentrevett for the feedback. I think at some point, we @cpuhrsch were talking about the option to reset the iterators. Since the current datasets are just iterators, it's pretty "cheap" to construct a new one once it has be depleted, especially the raw dataset files need to been downloaded once. But I think we should still think about the options @bentrevett proposed here, especially with `torchpipe`.

**Comment by parmeet** [comment]
This issue can be close as torchtext's datasets have already been migrated to TorchData datapipes. Refer to this issue https://github.com/pytorch/text/issues/1494 for more details.
